### DISCUSSION TITLE
netlab with trex traffic gen

### DISCUSSION THREAD

**Discussion by nagraj-kulkarni** [ipspace/netlab#1273] [Q&A] [answered]
Hello All,
Been trying few topologies with netlab and I must say its wonderful. Wondering if there is any way to hook up a traffic generator i.e trex tgen in the topology.  if not possible with the scripts, did anyone try manually to bring up another container and hook that with the netlab topology. Appreciate any ideas... Thanks..

**Comment by ssasso** [comment]
I know it's not exactly the same, but a starting point (that I personally use multiple times) is to use iperf3 from inside two Linux host containers.

I.e., on Alpine Linux containers you can install it using apk.

But keep in mind that the performances of a virtual NOS could be "not so good"...

**Comment by ipspace** [comment, marked as answer]
It's super easy to add anything that runs on Linux:

* Add a Linux VM or container (example with two Linux containers: https://github.com/ipspace/netlab/blob/dev/tests/integration/initial/01-interfaces.yml)
* If the software you want to run already comes with in a container, specify that container name as the device image (https://netlab.tools/nodes/#node-attributes). For example, use **image: trexcisco/trex** for your traffic generator (https://trex-tgn.cisco.com/trex/doc/trex_vm_manual.html#docker)
* Otherwise, use **ubuntu:22.04** container or **generic/ubuntu2204** Vagrant box and install software on it after the lab has started. You can also use custom configuration (https://netlab.tools/custom-config-templates/) to install the software when the lab is started, or build your own container image.

Adding network-control-plane software is a bit trickier, I'm planning to write a blog post describing how to add a BGP daemon, but you won't need that to get trex working.

However, as Stefano pointed out, testing performance of virtual network devices running in generic Docker/libvirt environment might not give you meaningful results.

**Reply by nagraj-kulkarni** [reply]
Great, Thank you Stefano and Ivan for the pointers and appreciate your quick responses. My intention is not to try any performance using trex. I just want to bring up topologies using netlab and have different kind of traffic (controlled) running across using trex.
bringing up trex as a container and use it as a device sounds good, let me check that with how to hook up the ports with other containers and i guess you already gave that pointer with two linux containers :)

**Comment by nagraj-kulkarni** [comment]
got Trex working as you suggested Ivan, working like a charm :)

**Reply by ssasso** [reply]
@nagraj-kulkarni It would be great if you can contribute to netlab documentation writing a small tutorial on how to run trex with netlab (i.e., installation and simple run) ;)

**Reply by nagraj-kulkarni** [reply]
Sure, will do

**Comment by ipspace** [comment]
Tutorial in https://netlab.tools/example/linux/
